programs/clearing_house/src/controller/position.rs

```diff
@@ -184,121 +207,179 @@ pub fn update_position_and_market(
-                .checked_sub(position.quote_asset_amount)
-                .ok_or_else(math_error!())?;
-            market.amm.quote_asset_amount_short = market
-                .amm
-                .quote_asset_amount_short
-                .checked_add(new_quote_asset_amount)
-                .ok_or_else(math_error!())?;
+                market.amm.quote_asset_amount_long = market
+                    .amm
+                    .quote_asset_amount_long
+                    .checked_add(new_quote_entry_amount)
+                    .ok_or_else(math_error!())?;
+                market.amm.quote_entry_amount_long = market
+                    .amm
+                    .quote_entry_amount_long
+                    .checked_add(new_quote_entry_amount)
+                    .ok_or_else(math_error!())?;
+            } else {
+                market.base_asset_amount_long = market
+                    .base_asset_amount_long
+                    .checked_sub(position.base_asset_amount)
+                    .ok_or_else(math_error!())?;
+                market.base_asset_amount_short = market
+                    .base_asset_amount_short
+                    .checked_add(new_base_asset_amount)
+                    .ok_or_else(math_error!())?;
+
+                market.amm.quote_asset_amount_long = market
+                    .amm
+                    .quote_asset_amount_long
+                    .checked_add(
+                        delta
+                            .quote_asset_amount
+                            .checked_sub(new_quote_entry_amount)
+                            .ok_or_else(math_error!())?,
+                    )
+                    .ok_or_else(math_error!())?;
+                market.amm.quote_entry_amount_long = market
+                    .amm
+                    .quote_entry_amount_long
+                    .checked_sub(position.quote_entry_amount)
+                    .ok_or_else(math_error!())?;
+                market.amm.quote_asset_amount_short = market
+                    .amm
+                    .quote_asset_amount_short
+                    .checked_add(new_quote_entry_amount)
+                    .ok_or_else(math_error!())?;
+                market.amm.quote_entry_amount_short = market
+                    .amm
+                    .quote_entry_amount_short
+                    .checked_add(new_quote_entry_amount)
+                    .ok_or_else(math_error!())?;
+            }
         }
     }
 
     // Update user position
-    if closed_position {
+    if let PositionUpdateType::Close = update_type {
         position.last_cumulative_funding_rate = 0;
         position.last_funding_rate_ts = 0;
-    } else if new_position || flipped_position {
+    } else if matches!(
+        update_type,
+        PositionUpdateType::Open | PositionUpdateType::Flip
+    ) {
         if new_base_asset_amount > 0 {
             position.last_cumulative_funding_rate = market.amm.cumulative_funding_rate_long;
         } else {
```

Comment:
i was thinking it might be good to have an assert statement here that position.last_funding = market.last_funding before we update the positions funding - idk if it would ever be hit but might be a good guardrail

Reply:
that's a good idea